synapse/lib/stormlib/imap.py

```diff
@@ -0,0 +1,376 @@
+
+@s_stormtypes.registry.registerLib
+class ImapLib(s_stormtypes.Lib):
+    '''
+    A Storm library to connect to an IMAP server.
+    '''
+    _storm_locals = (
+        {
+            'name': 'server',
+            'desc': '''
+            Open a connection to an IMAP server.
+
+            This method will wait for a "hello" response from the server
+            before returning storm:imap:server instance.
+            ''',
+            'type': {
+                'type': 'function', '_funcname': 'server',
+                'args': (
+                    {'type': 'str', 'name': 'host',
+                     'desc': 'The IMAP hostname.'},
+                    {'type': 'integer', 'name': 'port', 'default': 993,
+                     'desc': 'The IMAP server port.'},
+                    {'type': 'int', 'name': 'timeout', 'default': 30,
+                     'desc': 'The time to wait for all commands on the server to execute.'},
+                    {'type': 'bool', 'name': 'ssl', 'default': True,
+                     'desc': 'Use SSL to connect to the IMAP server.'},
+                ),
+                'returns': {
+                    'type': 'storm:imap:server',
+                    'desc': 'A new ``storm:imap:server`` instance.'
+                },
+            },
+        },
+    )
+    _storm_lib_path = ('inet', 'imap', )
+
+    def getObjLocals(self):
+        return {
+            'server': self.server,
+        }
+
+    async def server(self, host, port=993, timeout=30, ssl=True):
+        ssl = await s_stormtypes.tobool(ssl)
+        host = await s_stormtypes.tostr(host)
+        port = await s_stormtypes.toint(port)
+        timeout = await s_stormtypes.toint(timeout, noneok=True)
+
+        if ssl:
+            imap_cli = aioimaplib.IMAP4_SSL(host=host, port=port, timeout=timeout)
+        else:
+            imap_cli = aioimaplib.IMAP4(host=host, port=port, timeout=timeout)
+
+        async def fini():
+            # call protocol.logout() so fini() doesn't hang
+            await asyncio.wait_for(imap_cli.protocol.logout(), 5)
+
+        self.runt.snap.onfini(fini)
+
+        try:
+            await imap_cli.wait_hello_from_server()
+        except asyncio.TimeoutError:
+            raise s_exc.TimeOut(mesg='Timed out waiting for IMAP server hello.') from None
+
+        return ImapServer(self.runt, imap_cli)
+
+@s_stormtypes.registry.registerType
+class ImapServer(s_stormtypes.StormType):
+    '''
+    An IMAP server for retrieving email messages.
+    '''
+    _storm_locals = (
+        {
+            'name': 'list',
+            'desc': '''
+            List mailbox names.
+
+            By default this method uses a reference_name and pattern to return
+            all mailboxes from the root.
+            ''',
+            'type': {
+                'type': 'function', '_funcname': 'list',
+                'args': (
+                    {'type': 'str', 'name': 'reference_name', 'default': '""',
```

Comment:
Is the default the literal string `""` ?

Reply:
yup :)  https://datatracker.ietf.org/doc/html/rfc2060.html#page-30